### PR TITLE
Add support for Metadata Expressions including. tlsVersion.

### PR DESCRIPTION
**Pull request by jplevyak**
Signed-off-by: John Plevyak <jplevyak@gmail.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by jplevyak** [comment]
@mandarjog

File: api/wasm/cpp/proxy_wasm_enums.h [hunk 1]
```diff
@@ -15,15 +15,29 @@ enum class FilterDataStatus : EnumType {
...
+  Listener = 6,       // Immutable
+  Cluster = 7,        // Immutable
+  Expression = 8,     // The key is a string expression. Only proxy_getMetadata().
 };
+/*
+  Expression and their types:
+
+  request.protocol : string
+  response.protocol : string
+  request.destination_port : int32
```
Comment (kyessenov) [inline api/wasm/cpp/proxy_wasm_enums.h]: We should be careful not to define new names for things that are established in https://istio.io/docs/reference/config/policy-and-telemetry/attribute-vocabulary/ or https://cloud.google.com/iam/docs/conditions-overview. I'm happy to make another pass to fix the names in a follow-up

**Comment by kyessenov** [comment]
Just curious: why do we commit the large binary files? is there a CI check validating them?

File: api/wasm/cpp/proxy_wasm_enums.h (hunk 1 above)
Comment (mandarjog) [reply api/wasm/cpp/proxy_wasm_enums.h]: We should just use consistent names now, instead of another PR.

Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_enums.h]: I agree that we should have consistent names when we upstream, but this is an experimental repository and Kuat is going to replace all this code very soon.  This is just to get Mandar the last "must have" interface.  That said, if the correct names are obvious I'd like to make the change in this PR.

**Comment by jplevyak** [comment]
RE: the large binaries.  We have an issue to get emscripten working with bazel and then we can generate these files on the fly.  I agree that it is really annoying.  As soon as we get some cycles I am going to work on it.

File: api/wasm/cpp/proxy_wasm_enums.h (hunk 1 above)
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_enums.h]: The current names should not be consisdered part of the external interface.  We have the accessors for that, e.g. getRequestDestinationPort().  I can also prefix the current names with "_experimental_" or "_internal_" which will give a clear indication these are not part of a stable exernal interface.

**Comment by jplevyak** [comment]
Ping.

**Pull request merged**